Updated build variants and added publishing to evergreen

rozza opened the pull request:
Build variants are now: 

- jvm
- android - requires android sdk
- macos
- ios

Publishes snapshots on build to the snapshot repo: https://oss.sonatype.org/content/repositories/snapshots/org/mongodb/
Publishes releases to staging on tag.

KBSON-7

rozza committed:
Updated the build to support:

- jvm
- android - requires android sdk
- macos
- ios

KBSON-7

rozza committed:
Publish to maven central

Update the build to handle publishing.

KBSON-7

rozza commented:
A couple of questions:

Is it an issue to use the group: `org.mongodb.kbson` and then the namespace: `org.kbson` in the code?

Publishing makes a number of artifacts all namespaced in the group namespace `org.mongodb`:

* kbson-android-debug/ 
* kbson-android/
* kbson-iosarm64/
* kbson-iosx64/ 
* kbson-jvm/ 
* kbson-macos/ 
* kbson-macosarm64/
* kbson/

Should I make the group `org.mongodb.kbson` instead? It would be more discoverable / friendly when browsing. Is there a better group name to use?

jyemin commented:
> Is it an issue to use the group: org.mongodb.kbson and then the namespace: org.kbson in the code?

It's not a technical issue, but I think best practice would be to use `org.mongodb.kbson` for the package name as well.

rozza committed:
Update package / namespace to org.mongodb.kbson

cmelchior commented on build.gradle.kts:
```diff
@@ -31,61 +34,140 @@ plugins {
-            else -> throw GradleException("Host OS is not supported in Kotlin/Native.")
-        }
+    // Note: Android is configured separately below.
+    ios()
```
I think we need `iosX64`, `iosArm64` and `iosSimulatorArm64` as targets here.

At least looking at the targets `iosSimulatorArm64` is definitely missing.

cmelchior replied:
Also, it looks like we are using 1.7.10 for Kotlin, which prevents consuming this library below Kotlin 1.7.10 on Kotlin Native. It will give this warning:

```
 Task :library-sync:compileKotlinIosArm64 FAILED
w: skipping /Users/cm/.gradle/caches/modules-2/files-2.1/org.mongodb/kbson-iosarm64/1.0-SNAPSHOT/9809a26692a34e1f22ad1f3b9459494d43262f83/kbson-iosarm64-1.0-SNAPSHOT.klib. Incompatible abi version. The current default is '1.6.0', found '1.7.0'. The library produced by 1.7.10 compiler
e: Could not find "/Users/cm/.gradle/caches/modules-2/files-2.1/org.mongodb/kbson-iosarm64/1.0-SNAPSHOT/9809a26692a34e1f22ad1f3b9459494d43262f83/kbson-iosarm64-1.0-SNAPSHOT.klib" in [/Users/cm/Realm/realm-kotlin/packages/library-sync, /Users/cm/.konan/klib, /Users/cm/.konan/kotlin-native-prebuilt-macos-aarch64-1.6.10/klib/common, /Users/cm/.konan/kotlin-native-prebuilt-macos-aarch64-1.6.10/klib/platform/ios_arm64]
```

This is a problem as Realm Kotlin currently ships with minimum version support of 1.6.10.

JetBrains currently do not promise ABI stability between Kotlin Native releases, but you can consume libraries from earlier versions. This is why libraries should just try to go as low as possible.

The only case where it might cause issues is because of our Serialization dependency, which needs to use an older version (that is compatible with Kotlin 1.6.1), which might mean that certain bugs haven't been fixed. But it is worth a try IMO.

rozza replied:
Interesting does the `iosSimulatorArm64` version provide something different over the `iosArm64` version?

The reason I ask is if running the simulator is required for the library to build / test on macs

rozza replied:
Will look into the kotlin version - not sure if that will be just an easy fix or uncover some KMM issues.

Done!

cmelchior replied:
The reason we need the `iosSimulatorArm64` target is due to how Kotlin creates the XCode frameworks (as far as I understand). The details are here: https://youtrack.jetbrains.com/issue/KT-43667

It then also matters because Gradle will attempt to match the target people are running on to a specific dependency. Kotlin produces maven metadata to do this matching, which is why you are also seeing a of lof publications, and if we don't have publications for all targets people might not be able to compile their projects.

rozza replied:
Thanks @cmelchior will see if I can get evergreen to publish that variant as well.

rozza replied:
This should now be done :)

cmelchior commented on build.gradle.kts:
```diff
@@ -31,61 +34,140 @@ plugins {
+
+    @Suppress("UnstableApiUsage")
+    configure<com.android.build.gradle.LibraryExtension> {
+        namespace = "org.kbson"
```
```suggestion
        namespace = "org.mongodb.kbson"
```

rozza replied:
Done - good catch

cmelchior commented on build.gradle.kts:
```diff
@@ -16,13 +16,16 @@
-group = "org.mongodb"
+group = "org.mongodb.kbson"
 
 version = "1.0-SNAPSHOT"
```
```suggestion
version = "1.0.0-SNAPSHOT"
```

So it matches full semantic versioning

rozza replied:
Done - good catch

cmelchior reviewed (changes requested):
Nice. I was able to start trying to integrate it. I ran into a few problems though which we need to fix.

rozza committed:
Fix namespace and version

rozza committed:
Target Kotlin 1.6.10+

rozza commented:
Published to: https://oss.sonatype.org/content/repositories/snapshots/org/mongodb/kbson/

cmelchior commented:
Hmm, something seems to be off with the metadata. At least I'm getting this when trying to consume the artifacts:

```
         > The consumer was configured to find a usage of 'kotlin-api' of a library, preferably optimized for non-jvm, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'common'. However we cannot choose between the following variants of org.mongodb.kbson:kbson:1.0.0-SNAPSHOT:20221004.102555-1:
             - debugApiElements-published
             - debugRuntimeElements-published
             - iosArm64ApiElements-published
             - iosX64ApiElements-published
             - jvmApiElements-published
             - jvmRuntimeElements-published
             - macosApiElements-published
             - macosArm64ApiElements-published
             - releaseApiElements-published
             - releaseRuntimeElements-published
           All of them match the consumer attributes:
             - Variant 'debugApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares an API of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
                 - Unmatched attributes:
                     - Provides attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
             - Variant 'debugRuntimeElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a runtime of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
                 - Unmatched attributes:
                     - Provides attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
             - Variant 'iosArm64ApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a usage of 'kotlin-api' of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Unmatched attributes:
                     - Provides attribute 'artifactType' with value 'org.jetbrains.kotlin.klib' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
                     - Provides attribute 'org.jetbrains.kotlin.native.target' with value 'ios_arm64' but the consumer didn't ask for it
             - Variant 'iosX64ApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a usage of 'kotlin-api' of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Unmatched attributes:
                     - Provides attribute 'artifactType' with value 'org.jetbrains.kotlin.klib' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
                     - Provides attribute 'org.jetbrains.kotlin.native.target' with value 'ios_x64' but the consumer didn't ask for it
             - Variant 'jvmApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares an API of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm':
                 - Unmatched attributes:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides its elements packaged as a jar but the consumer didn't ask for it
                     - Provides integration status but the consumer didn't ask for it
             - Variant 'jvmRuntimeElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a runtime of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm':
                 - Unmatched attributes:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides its elements packaged as a jar but the consumer didn't ask for it
                     - Provides integration status but the consumer didn't ask for it
             - Variant 'macosApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a usage of 'kotlin-api' of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Unmatched attributes:
                     - Provides attribute 'artifactType' with value 'org.jetbrains.kotlin.klib' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
                     - Provides attribute 'org.jetbrains.kotlin.native.target' with value 'macos_x64' but the consumer didn't ask for it
             - Variant 'macosArm64ApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a usage of 'kotlin-api' of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Unmatched attributes:
                     - Provides attribute 'artifactType' with value 'org.jetbrains.kotlin.klib' but the consumer didn't ask for it
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
                     - Provides attribute 'org.jetbrains.kotlin.native.target' with value 'macos_arm64' but the consumer didn't ask for it
             - Variant 'releaseApiElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares an API of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
                 - Unmatched attributes:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
             - Variant 'releaseRuntimeElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a runtime of a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
                 - Unmatched attributes:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
                     - Provides integration status but the consumer didn't ask for it
           The following variants were also considered but didn't match the requested attributes:
             - Variant 'iosArm64MetadataElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Incompatible because this component declares a usage of 'kotlin-metadata' of a component and the consumer needed a usage of 'kotlin-api' of a component
                 - Other compatible attribute:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
             - Variant 'iosX64MetadataElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Incompatible because this component declares a usage of 'kotlin-metadata' of a component and the consumer needed a usage of 'kotlin-api' of a component
                 - Other compatible attribute:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
             - Variant 'macosArm64MetadataElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Incompatible because this component declares a usage of 'kotlin-metadata' of a component and the consumer needed a usage of 'kotlin-api' of a component
                 - Other compatible attribute:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
             - Variant 'macosMetadataElements-published' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native':
                 - Incompatible because this component declares a usage of 'kotlin-metadata' of a component and the consumer needed a usage of 'kotlin-api' of a component
                 - Other compatible attribute:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)
             - Variant 'metadataApiElements' capability org.mongodb.kbson:kbson:1.0.0-SNAPSHOT declares a library, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'common':
                 - Incompatible because this component declares a usage of 'kotlin-metadata' of a component and the consumer needed a usage of 'kotlin-api' of a component
                 - Other compatible attribute:
                     - Doesn't say anything about its target Java environment (preferred optimized for non-jvm)

rozza commented:
@cmelchior how might I replicate that error? Seems similar to https://youtrack.jetbrains.com/issue/KT-48370 and https://youtrack.jetbrains.com/issue/KT-50263

cmelchior commented:
I was using this PR https://github.com/realm/realm-kotlin/pull/1045
And then calling `/realm-kotlin/packages/gradlew publishAllPublicationsToTestRepository`

I assume it would also be reproducible by a simple KMP project, but I haven't tried it.

jyemin commented:
I don't think my review here is going to be helpful.  Removing myself.

cmelchior commented:
I was trying to reproduce in a complete new KMP app (created from the Android Studio template). It seemed to compile fine when disabling iosSimulator target, so the above error might be something in Realm Kotlin. I will need to investigate that more.

But it also looks like sources are not being packaged correctly. At least jumping to the BsonDocument definition gave me this:

![image](https://user-images.githubusercontent.com/406066/194004506-81d33da9-481e-4bb8-a791-ede4ef6a6cb7.png)

rozza committed:
Fix metadata hierarchical project structure

Updated gradle.properties to support kotlin 1.6.10
https://kotlinlang.org/docs/multiplatform-hierarchy.html

rozza commented:
@cmelchior I think the variant issue has been fixed in 07ecadc - that also fixed the code inspection view in intellij for me.

A new snapshot has been uploaded to sonatype.

rozza committed:
Remove resources plugin and implement natively

rozza committed:
Fix Decimal128 with Kotlin 1.6.10

Named regexes are not supported on native until kotlin 1.7.x
https://youtrack.jetbrains.com/issue/KT-41890/Support-named-capture-groups-in-Regex-on-Native

rozza committed:
Update android lint baseline now using 1.6.10

cmelchior commented on src/macosTest/kotlin/org/mongodb/kbson/corpus/ResourceLoader.kt:
```diff
@@ -0,0 +1,35 @@
+internal actual object ResourceLoader {
+
+    actual fun readText(resourceName: String): String {
+        val file: CPointer<FILE>? = fopen("./src/commonTest/resources/$resourceName", "r")
```
Don't you need to close this file again?

rozza replied:
I have updated how I get the resource.

cmelchior commented on src/commonTest/kotlin/org/mongodb/kbson/BsonArrayTest.kt:
```diff
@@ -27,17 +27,17 @@ import kotlin.test.assertTrue
 
     private val listOfBsonValues = listOf(BsonBoolean.TRUE, BsonBoolean.FALSE)
-    private val bsonValue = BsonArray(listOfBsonValues)
+    private val bsonValue = org.mongodb.kbson.BsonArray(listOfBsonValues)
```
Any reason to include the package name here?

rozza replied:
Missed that one - a hangover from the automatic refactoring moving the package.

cmelchior commented on src/commonMain/kotlin/org/mongodb/kbson/BsonValue.kt:
```diff
@@ -40,7 +40,7 @@ public sealed class BsonValue {
     /** @return true if this is a BsonArray, false otherwise */
     public fun isArray(): Boolean {
-        return this is BsonArray
+        return this is org.mongodb.kbson.BsonArray
```
I don't think we need to include the package name here? and the rest of this class

rozza replied:
Done

cmelchior reviewed (commented):
Nice. This seems to fix most of the issues I was hitting. It looks like the iosSimulatorArm64 target is still missing though:

![image](https://user-images.githubusercontent.com/406066/194120561-638bcdd0-b464-451c-b534-11d92ad30bc0.png)

rozza committed:
Update build - move snapshot release to top level

rozza committed:
Add support for iosSimulatorArm64

rozza committed:
Update missed paths from auto refacoring in intellij (org.kbson -> org.mongodb.kbson)

rozza committed:
Ensure android tests can find resources and fix Corpus test to throw on missing resources

rozza committed:
Copy macosArm64 resources

rozza committed:
Added section headers to build.gradle.kts

rozza commented:
Right - this should now be ready for review again.

* iosSimulatorArm64 support added, 
  Had to remove the kotlinx-resources plugin as adding iosSimulatorArm64, caused the variant errors seen previously.
  So there are mulitple copy resources tasks / helpers required until support for resources is added to KMM.
* Decimal128 - had to be updated as 1.6.10 doesn't support names in regexes and it failed on macos
* Updated the build (evergreen) so that publish snapshot and publish release are both top level in the matrix.
* Added [KBSON-12](https://jira.mongodb.org/browse/KBSON-12) to understand why the build passes when platform specific tests failed. Both Decimal128 and corpus tests reported failures on macosArm64 but the build as a whole returned as Successful.

cmelchior reviewed (approved):
Nice 🚀 . My sample project is now building on all targets and the source lookup looks correct

rozza commented:
Manually rebased and merged
